File: src/main/scala/kim/dencka/scala/easy/other/GreatestCommonDivisorOfStrings.scala
package kim.dencka.scala.easy.other

object GreatestCommonDivisorOfStrings extends App {

  // solution: https://leetcode.com/problems/greatest-common-divisor-of-strings/solution/
  def gcdOfStrings(str1: String, str2: String): String =
    if (str1 + str2 != str2 + str1) ""
    else str1.substring(0, gcd(str1.length, str2.length))

  @annotation.tailrec
  def gcd(a: Int, b: Int): Int = if (b == 0) a else gcd(b, a % b)

  //  // brute force
//  def gcdOfStrings(str1: String, str2: String): String =
//    if (str1.length > str2.length) findGDC(str1, str2, str2)
//    else findGDC(str2, str1, str1)
//
//  @annotation.tailrec
//  def findGDC(long: String, short: String, acc: String): String =
//    if ((long.split(acc).isEmpty && short.split(acc).isEmpty) || acc.isEmpty) acc
//    else findGDC(long, short, acc.init)

  println(gcdOfStrings("LEET", "CODE")) // output: ""
  println(                              // output: TAUXX
    gcdOfStrings("TAUXXTAUXXTAUXXTAUXXTAUXX", "TAUXXTAUXXTAUXXTAUXXTAUXXTAUXXTAUXXTAUXXTAUXX")
  )
}

File: src/main/scala/kim/dencka/java/easy/GreatestCommonDivisorOfStrings.java
package kim.dencka.java.easy;

public class GreatestCommonDivisorOfStrings {
    public static void main(String[] args) {
        String str1 = "ABCABC", str2 = "ABC";
        System.out.println("#1: " + gcdOfStrings(str1, str2));

        String str3 = "ABABAB", str4 = "ABAB";
        System.out.println("#2: " + gcdOfStrings(str3, str4));

        String str5 = "LEET", str6 = "CODE";
        System.out.println("#3: " + gcdOfStrings(str5, str6));
    }

    public static String gcdOfStrings(String str1, String str2) {
        if (!(str1 + str2).equals(str2 + str1)) return "";

        int gcdLength = gcd(str1.length(), str2.length());
        return str1.substring(0, gcdLength);
    }

    public static int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }
}
